docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,258 @@
+management areas of the code.
+
+In order to support replacement of a component, a new public function
+will be added to the public interface of `node` with a signature that
+resembles the following:
+
+```go
+func NewWithServices(conf *config.Config,
+	logger log.Logger,
+	cf proxy.ClientCreator,
+	gen *types.GenesisDoc,
+	srvs []service.Service,
+) (service.Service, error) {
+```
+
+The `service.Service` objects will be initialized in the order their
+supplied, after all pre-configured/default services have started (and
+shut down first, comparatively.) If any of the services implement
+additional interfaces, that allow them to replace specific default
+services. `NewWithServices` will validate input service lists with the
+following rules:
+
+- no running services
+- for replacing default services, callers cannot supply more than
+  one replacement reactor for any specified type.
+
+If callers violate any of these rules, `NewWithServices` will return
+an error.
+
+## Consequences
+
+### Positive
+
+- The node package will become easier to maintain.
+
+- It will become easier to add additional services within tendermint
+  nodes.
+
+- It will become possible to replace default components in the node
+  package without vendoring the tendermint repo and modifying internal
+  code.
+
+- The current end-to-end (e2e) test suite will be able to prevent any
+  regressions, and the new functionality can be thoroughly unittested.
+
+- The scope of this project is very narrow, which minimizes risk and
+
+### Negative
+
+- This increases our reliance on the `service.Service` interface which
+  is probably not an interface that we want to fully commit to.
+
+- This proposal implements a fairly minimal set of functionality and
+  leaves open the possibility for many additional features which are
+  not included in the scope of this proposal.
+
+### Neutral
+
+N/A
+
+## Open Questions
+
+- To what extent does this new initialization framework need to accommodate
+  the legacy p2p stack? Would it be possible to delay a great deal of this
+  work to the 0.36 cycle to avoid this complexity?
+
+  - Answer: _depends on timing_, and the requirement to ship pluggable
+	reactors in 0.35.
+
+- Where should additional public types be exported for the 0.35
+  release?
+
+  Related to the general project of API stabilization there is a
+  desire to deprecate the `types` package, and move these public types
+  into a new `pkg` hierarchy; however, the design of the `pkg`
+  interface is currently underspecified. If `types` is going to remain
+  for the 0.35 release, then we should consider the impact of using
+  multiple organizing modalities for this code within a single release.
+
+## Future Work
+
+- Improve or simplify the `service.Service` interface. There are some
+  pretty clear limitations with this interface as written (there's no
+  way to timeout slow startup or shut down, the cycle between the
```

Comment:
This could be a good target for context plumbing work. I think I've seen at least one issue touching on that, and we could potentially address it as part of a wider (but separate) context plumbing update. That's something that's probably worth doing before 1.0, but could wait till after this more immediate step.

Reply:
Yeah, I totally agree. I definitely want to do context plumbing, sometime later this year, but I don't think this proposal is exactly the right place to do it? It's clearly related but I think drawing lines between these kinds of projects is useful.